rasa/nlu/classifiers/diet_classifier.py

```diff
@@ -658,125 +669,226 @@ def _use_default_label_features(self, label_ids: np.ndarray) -> List[FeatureArra
             )
 
-        if (
-            label_attribute
-            and model_data.does_feature_not_exist(label_attribute, SENTENCE)
-            and model_data.does_feature_not_exist(label_attribute, SEQUENCE)
-        ):
+        # (2) add feature for the labels and entities if in training mode
+        if training:
+
+            labels_to_consider = []
+            if self.label_attribute is not None:
+                labels_to_consider.append(self.label_attribute)
+            if self.component_config[ENTITY_RECOGNITION] and self._entity_tag_specs:
+                # Add entities as labels only during training and only if there was
+                # training data added for entities with DIET configured to predict
+                # entities.
+                labels_to_consider.append(ENTITIES)
+
+            (
+                features_for_labels,
+                sparse_feature_sizes_for_labels,
+            ) = model_data_utils.featurize_training_examples(
+                messages,
+                attributes=labels_to_consider,
+                entity_tag_specs=self._entity_tag_specs,
+                featurizers=self.component_config[FEATURIZERS],
+                bilou_tagging=self.component_config[BILOU_FLAG],
+            )
+
+            # add the label features to the text features
+            for feats_txt, feats_lbl in zip(features, features_for_labels):
+                feats_txt.update(feats_lbl)
+            sparse_feature_sizes.update(sparse_feature_sizes_for_labels)
+
+        return features, sparse_feature_sizes
+
+    def _fallback_to_default_label_fatures_if_necessary(
+        self, label_ids: List[int], model_data: RasaModelData,
+    ) -> None:
+        """Fills in default label features for the label attribute if necessary.
+
+        Here, "required" means that no features are present for the label attribute
+        in the given model_data.
+        Note that if there is no label_attribute set, then nothing will be changed.
+
+        Args:
+          label_ids: list of label ids (indices) that contains entries for just as
+            many records as the `model_data`
+          model_data: the `RasaModelData` which possibly does not contain sequence
+            or sentence features for the label attribute
+        """
+        if self.label_attribute is None:
+            return
+
+        # If *no* label features have been found before, then we load the default
+        # label features that should've been computed by an `_create_label_data`.
+        sentence_missing = model_data.does_feature_not_exist(
+            self.label_attribute, SENTENCE
+        )
+        sequence_missing = model_data.does_feature_not_exist(
+            self.label_attribute, SEQUENCE
+        )
+        needs_sentence = self._needs_sentence_features_for_labels()
+        if sentence_missing and (needs_sentence or sequence_missing):
+            if (not sequence_missing) and sentence_missing and needs_sentence:
```

Comment:
first if:  if the sentence is missing, but we really need the sentence or (the sentence is still missing and) sequence is missing as well, then we're in trouble...
second if: special case where the sequence is *not* missing but "the sentence is missing, but we really need the sentence"

Reply:
Sorry, had overlooked that you wanted that as a comment in there. Do you really think this is required given my comment would look pretty much 1:1 like the boolean expression (except the bit of distributed law in it)? Any suggestion for a short comment that would help?

Reply:
Ha, now that you put it in words seems fine to me, but I think there's some redundancy which complicates it... Firstly, you don't need `sentence_missing` in the second `if` because control will never come there if that wasn't True in the first place. So, can we write the second `if` as:

```
if needs_sentence and not sequence_missing:
```
?

Reply:
This was meant to be better readable but apparently just confused then. ...  Sure that I should remove it because the redundant version is actually close to the explanation above?

Reply:
done